Add Delphi 10.2 Tokyo support via dedicated client and tool frame

andrevfarias opened the pull request:
This PR introduces initial compatibility for Delphi 10.2 Tokyo by adding a dedicated client build and a Tokyo-specific tool window implementation. It addresses the request to support older Delphi IDEs, motivated by issue [#14](https://github.com/integrated-application-development/delphilint/issues/14).

Summary of changes
- New Tokyo tool frame:
  - Added `DelphiLint.ToolFrame.Tokyo.pas/.dfm`, using `TListView` and `TWebBrowser` (SHDocVw) instead of newer components that are not available on 10.2.
- New 10.2 Tokyo client projects:
  - Added `client/source/DelphiLintClient250.dpk`, `DelphiLintClient250.dproj`, and group project `client/DelphiLintClientProjects250.groupproj`.
  - Added test project `client/test/DelphiLintClientTest250.dpr/.dproj`.
- Conditional compilation for older IDEs:
  - `DelphiLint.Plugin.pas` selects `DelphiLint.ToolFrame.Tokyo` when `CompilerVersion < 33.0`.
  - `DelphiLint.IDEContext.pas` uses `IOTAIDEThemingServices250` where required.
- Build and packaging:
  - `scripts/build.ps1` updated to support version `250` (Tokyo), map registry version `19.0`, and use `DCC_UseMSBuildExternally=true` to avoid long command-line issues.
- Minor compatibility tweaks across units (Analyzer, HtmlGen, Server, Settings, etc.) to compile and run on Tokyo.
- No behavior changes for Alexandria/Athens; newer IDEs continue using the existing frame and WebView2.

Why
- Many teams still use Delphi 10.2 Tokyo, and the main client relies on components not available there. This PR provides a safe, isolated implementation for Tokyo without impacting newer Delphi versions. See [issue #14](https://github.com/integrated-application-development/delphilint/issues/14).

Build notes
- Use the existing build pipeline with the 250 target (e.g., `build.ps1 250`).

Impact
- Adds 10.2 support with a separate UI implementation; maintains current behavior for Delphi 11/12.

andrevfarias committed:
refactor: add language compatibity for legacy delphi version compilers

andrevfarias committed:
feat: Add Delphi project for Tokyo compatibility

fourls commented on client/source/DelphiLint.FileLogger.pas:
```diff
@@ -66,7 +66,7 @@ constructor TFileLogger.Create(LogPath: string);
 
   if not TFile.Exists(FLogPath) then begin
-    FreeAndNil(TFile.Create(FLogPath));
+    TFile.Create(FLogPath).Free;
```
What's the purpose of these changes?

andrevfarias replied:
In older Delphi versions (like Tokyo), passing a constant object to `FreeAndNil()` as a var parameter triggers the compiler error `E2197 Constant object cannot be passed as var parameter`. Since this is a local instantiation and we don't strictly need to nil the reference, calling `.Free` directly solves the issue.

fourls commented on client/source/DelphiLint.Handlers.pas:
```diff
@@ -319,9 +319,9 @@ destructor TEditorHandler.Destroy;
+    Hook.Free;
   end;
-  FreeAndNil(FPopupHooks);
+  FreeAndNil(FPopupHooks); // Remover?
```
What does this comment mean?

andrevfarias replied:
Just garbage, I've removed it.

fourls commented on client/source/DelphiLint.HtmlGen.pas:
```diff
@@ -354,6 +354,59 @@ function TRuleHtmlGenerator.GenerateRuleDescriptionHtml(Description: TRuleDescri
 
 function TRuleHtmlGenerator.BuildHtmlPage(BodyHtml: string; BodyClass: string): string;
+
+  function GetLegacyScript: string;
```
Webpack/Babel can polyfill for old IE versions (see `client/jslib/webpack.config.js`) so there's no need to include it ourselves. Could you identify the IE version you want to support and try configuring webpack to bundle for that version?

andrevfarias replied:
The TWebBrowser uses an IE7 engine, and even when configuring Webpack / Babel to target that version, the runtime still lacks fundamental features (e.g., Object.defineProperty). These cannot be reliably polyfilled, which leads to script errors at execution time.

fourls commented on client/source/DelphiLint.IDEContext.pas:
```diff
@@ -359,7 +359,11 @@ procedure TToolsApiServices.ApplyTheme(Component: TComponent);
 
 procedure TToolsApiServices.RegisterFormClass(FormClass: TCustomFormClass);
 begin
+{$IF CompilerVersion < 33.0}
+  (BorlandIDEServices as IOTAIDEThemingServices250).RegisterFormClass(FormClass);
+{$ELSE}
   (BorlandIDEServices as IOTAIDEThemingServices).RegisterFormClass(FormClass);
+{$ENDIF}
```
There's no need to do the ifdef here, just change to 250 if it's a compatibility issue. Newer Delphi versions also have the 250 version.

andrevfarias replied:
Done.

fourls commented on client/source/DelphiLint.OptionsForm.pas:
```diff
@@ -188,7 +188,7 @@ procedure TLintOptionsForm.RefreshTheme;
 begin
   LintContext.IDEServices.ApplyTheme(Self);
-  WindowColor := StyleServices(Self).GetSystemColor(clWindow);
+  WindowColor := StyleServices.GetSystemColor(clWindow);  
```
Please remove this trailing whitespace.

andrevfarias replied:
Done.

fourls commented on client/source/DelphiLint.Server.pas:
```diff
@@ -1027,7 +1032,11 @@ procedure TLintServerTcpConnection.DoSendMessage(Msg: TTaggedMessage);
   Message := Msg.Extract;
 
   if Assigned(Message.Data) then begin
+{$IF CompilerVersion < 33.0}
+    DataBytes := TEncoding.UTF8.GetBytes(EscapedString(Message.Data.ToString));
+{$ELSE}
     DataBytes := TEncoding.UTF8.GetBytes(Message.Data.ToString);
+{$ENDIF}
```
What's the reasoning behind this change?

andrevfarias replied:
In older Delphi versions (prior to 10.3 Rio), System.JSON has a known bug where TJSONValue.ToString() fails to properly escape control characters like newlines (\n) and tabs (\t), which results in invalid JSON. This change manually escapes those characters to ensure the payload is parsed correctly by the server.

fourls commented on client/source/DelphiLint.Server.pas:
```diff
@@ -1123,4 +1132,25 @@ function TTaggedMessage.Extract: TLintMessage;
+  if AValue.StartsWith('"') then
+    AValue := AValue.Substring(1, AValue.Length-2);
+
+  // Tratar estes caracteres especiais: \, ", #$8, #$9, #$a, #$c, #$d
```
Please translate this comment into English.

andrevfarias replied:
Done.

fourls commented on client/source/DelphiLint.Settings.pas:
```diff
@@ -215,16 +215,16 @@ procedure TLintSettings.SyncTokenMap;
   FTokensMap.Clear;
 
-  HostTokenPairs := SplitString(SonarHostTokens, ',');
+  HostTokenPairs := TArray<string>(SplitString(SonarHostTokens, ','));
```
What's the purpose of these changes?

andrevfarias replied:
Implicit casting was causing compiling error on older delphi versions.

fourls commented on client/source/DelphiLint.SettingsFrame.dfm:
Are these changes accidental?

andrevfarias replied:
These properties aren't compatible with delphi 10.2, so I removed it, maybe it can cause some minor visual issues at RadioGroup header fonts and outline. I couldn't find a better solution yet.

fourls commented on client/source/DelphiLint.SetupForm.pas:
```diff
@@ -162,7 +162,7 @@ procedure TLintSetupForm.RefreshTheme;
 begin
   LintContext.IDEServices.ApplyTheme(Self);
-  WindowColor := StyleServices(Self).GetSystemColor(clWindow);
+  WindowColor := StyleServices.GetSystemColor(clWindow);  
```
Please remove this trailing whitespace.

andrevfarias replied:
Done.

fourls commented on client/source/DelphiLint.ToolFrame.Tokyo.pas:
Please

- Translate comments to English
- Remove Tokyo-specific references
- Include this unit in all DelphiLintClient dprojs

andrevfarias replied:
Partially done.
As DelphiLint.ToolFrame.Tokyo.pas uses the same component names as DelphiLint.ToolFrame.pas, keep both included in the projects causes conflicts from duplicated classes, I've kept only in DelphiLintClient250 project.

fourls commented on scripts/build.ps1:
Please revert this change. As stated previously, we will not be officially supporting building for versions of Delphi before Delphi 11, especially not specifically for 10.2.

andrevfarias replied:
Understood. I have reverted the changes made to the build script (`build.ps1`).

fourls commented on scripts/TEMPLATE_install.ps1:
What's the purpose of this change?

andrevfarias replied:
That change was a leftover from local testing and was unnecessary for the scope of this PR. I have already reverted it.

fourls commented on server/delphilint-server/src/main/java/au/com/integradev/delphilint/analysis/AnalysisOrchestrator.java:
```diff
@@ -107,7 +107,9 @@ private static AnalysisConfiguration buildConfiguration(
 
   private static Charset getConfiguredCharset(Map<String, String> properties) {
-    if (properties.containsKey("sonar.encoding")) {
+    if (properties.containsKey("sonar.sourceEncoding")) {
+      return Charset.forName(properties.get("sonar.sourceEncoding"));
+    } else if (properties.containsKey("sonar.encoding")) {
```
Oops, this seems like a legitimate bug, thanks for picking this up. Could you move this out into a separate commit?

andrevfarias replied:
Done. I have split the history and isolated this fix into its own dedicated commit (`fix: update charset configuration key for compatibility`).

fourls reviewed (changes requested):
Hi @andrevfarias, thank you for the PR! It mostly looks good and I'm happy to accept most of its changes to improve compatibility with older Delphi versions. 

I have made some comments. Because I do not have Tokyo installed, I will not be able to test it or ensure any quality guarantees. 

> Many teams still use Delphi 10.2 Tokyo, and the main client relies on components not available there.

I'm not super comfortable with how specific this change is to 10.2. Many teams still use Delphi 7, XE3, etc, but that doesn't seem like a reason to require a separate tool frame for each of those versions as well. Could we reframe this as a more general "legacy tool frame" instead of something specific to a minor release of the compiler?

Also, the build script should not be updated to build unsupported versions of Delphi. The purpose of the build script is for packaging and release, which won't need to be done for unsupported versions. There's also nothing special about 10.2 that necessitates it being added to the build script, and it implies a guarantee that it will be compatible (which there is not).

The script doesn't do anything fancy, so if you'd like to compile an unsupported version's BPL it can just be compiled in the IDE / via MSBuild / whatever approach is best for your setup. Or by making the changes to the build script locally.

andrevfarias committed:
chore: Resolve themed style not applyed when using older Delphi versions

andrevfarias committed:
fix: Update issue creation date handling with regex for better ISO8601 compliance

andrevfarias committed:
refactor: force TArray<string> casting to avoid Legacy Delphi compiling errors

- Older Delphi versions throw implicit casting error when using implicit
  casting of TArray<string> on function return types.
- This change forces the casting to avoid the implicit casting error.

andrevfarias committed:
refactor: replace FreeAndNil by .Free to avoid Legacy Delphi compiling errors

- Older Delphi versions throw the error:
  E2197 Constant object cannot be passed as var parameter
  when a constant object is passed as var parameter.
- When a local var is assigned to nil, it's not necessary to call
  FreeAndNil, .Free is enough.

andrevfarias committed:
refactor: call StyleServices passing `Self` parameter for Delphi Legacy compatibility

- Older Delphi versions requires to pass `Self` parameter when
  calling StyleServices.
- Newer Delphi versions also have this overloaded method.

andrevfarias committed:
refactor: use IOTAIDEThemingServices250 for Delphi Legacy compatibility

andrevfarias committed:
refactor: update JSON handling for Delphi Legacy compatibility

- Older Delphi versions does not accept direct item access
  via `Json[Index]`. Used `Json.Items[Index]` instead.
- Older Delphi versions doesn't have `TJsonValue.AsType<string>`.
  Used explicit casting to TJsonString instead.
- Use ParseJSONValue from TJSONObject instead of TJSONValue.

andrevfarias committed:
refactor: update destructor loop for Delphi Legacy compatibility

- Replaced the for-in loop with an index-based for loop to fix the
  E2197 compiler error: 'Constant object cannot be passed as var
  parameter' on freeAndNil call when compiling in older Delphi versions.
  This change ensures compatibility with legacy Delphi versions.

andrevfarias committed:
refactor: add EscapedString function for Delphi Legacy compatibility

- Introduced the EscapedString function to handle special character
  escaping for older Delphi versions.
- Updated data handling in the server to use EscapedString when
  CompilerVersion is less than 33.0.

andrevfarias committed:
fix: update charset configuration key for compatibility

- Changed the property key from `sonar.encoding` to `sonar.sourceEncoding`
  to align with updated configuration standards.

andrevfarias committed:
refactor: enhance HTML generation for legacy compatibility

- Implemented legacy script handling in HTML generation to ensure
  compatibility with older browsers.

andrevfarias committed:
refactor: remove .dfm properties incompatible with legacy Delphi versions

andrevfarias committed:
feat: introduce DelphiLintClient250 project and enhance HTML generation

- Added new project files for DelphiLintClient250.
- Implemented ToolFrame.Legacy to support legacy Delphi versions while
  maintaining functionality.

andrevfarias committed:
feat: add DUnitX test project, test group, and supporting mock/utility files for client validation

andrevfarias commented:
Hi @fourls, thanks for the feedback! 
I have rebased the entire branch and cleaned up the commit history to address all your requests. 

Here is a summary of the updates:
1. Reframed the specific "10.2 Tokyo" implementation into a broader "Legacy" compatibility approach.
2. Reverted all changes to the build/install scripts (`build.ps1` and `TEMPLATE_install.ps1`).
3. Isolated the legitimate bug fix (Charset configuration in `AnalysisOrchestrator.java`) into its own separate commit.
4. Split the changes into smaller, logical commits to make future reviews and history tracking easier.

Let me know if everything looks good now!
